transports/webrtc-websys/src/stream.rs

```diff
@@ -0,0 +1,339 @@
+    state::{Closing, State},
+    MAX_DATA_LEN,
+};
+use send_wrapper::SendWrapper;
+use std::io;
+use std::pin::Pin;
+use std::task::{ready, Context, Poll};
+use web_sys::{
+    RtcDataChannel, RtcDataChannelInit, RtcDataChannelState, RtcDataChannelType, RtcPeerConnection,
+};
+
+mod drop_listener;
+mod framed_dc;
+mod poll_data_channel;
+
+pub(crate) use drop_listener::DropListener;
+
+// Max message size that can be sent to the DataChannel
+const MAX_MESSAGE_SIZE: u32 = 16 * 1024;
+
+// How much can be buffered to the DataChannel at once
+const MAX_BUFFERED_AMOUNT: u32 = 16 * 1024 * 1024;
+
+// How long time we wait for the 'bufferedamountlow' event to be emitted
+const BUFFERED_AMOUNT_LOW_TIMEOUT: u32 = 30 * 1000;
+
+/// The Browser Default is Blob, so we must set ours to Arraybuffer explicitly
+const ARRAY_BUFFER_BINARY_TYPE: RtcDataChannelType = RtcDataChannelType::Arraybuffer;
+
+/// Builder for DataChannel
+#[derive(Default, Debug)]
+pub struct DataChannelConfig {
+    negotiated: bool,
+    id: u16,
+}
+
+/// DataChannel type to ensure only the properly configurations
+/// are used when building a WebRTCStream
+pub enum DataChannel {
+    Regular(RtcDataChannel),
+    Handshake(RtcDataChannel),
+}
+
+impl DataChannel {
+    pub fn new_regular(peer_connection: &RtcPeerConnection) -> Self {
+        Self::Regular(DataChannelConfig::create(peer_connection))
+    }
+
+    pub fn new_handshake(peer_connection: &RtcPeerConnection) -> Self {
+        Self::Handshake(DataChannelConfig::create_handshake_channel(peer_connection))
+    }
+}
+
+/// From DataChannel enum to RtcDataChannel
+impl From<DataChannel> for RtcDataChannel {
+    fn from(dc: DataChannel) -> Self {
+        match dc {
+            DataChannel::Regular(dc) => dc,
+            DataChannel::Handshake(dc) => dc,
+        }
+    }
+}
+
+/// Builds a Data Channel with selected options and given peer connection
+///
+/// The default config is used in most cases, except when negotiating a Noise handshake
+impl DataChannelConfig {
+    /// Create a new [DataChannelConfig] with the default values
+    /// Build the [RtcDataChannel] on a given peer connection
+    /// by calling
+    ///
+    /// ```no_run
+    /// let channel = DataChannelConfig::new().create_from(peer_conn);
+    /// ```
+    fn create(peer_connection: &RtcPeerConnection) -> RtcDataChannel {
+        Self::default().create_from(peer_connection)
+    }
+
+    /// Creates a new data channel with Handshake config,
+    /// uses negotiated: true, id: 0
+    /// Build the RtcDataChannel on a given peer connection
+    /// by calling `create_from(peer_conn)`
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// let channel = DataChannelConfig::new_handshake_channel(&peer_connection)
+    /// ```
+    fn create_handshake_channel(peer_connection: &RtcPeerConnection) -> RtcDataChannel {
+        Self::default()
+            .negotiated(true)
+            .id(0)
+            .create_from(peer_connection)
+    }
```

Comment:
Would it make sense to move this into the `upgrade` module given that it is unrelated to the regular `Stream`?

Reply:
Yes it would make sense, that seems to be the only place it's used. I'll extract it.